flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/FlinkDistribution.java

```diff
@@ -0,0 +1,219 @@
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * A wrapper around a Flink distribution.
+ */
+public final class FlinkDistribution extends ExternalResource {
+
+	private static final Logger LOG = LoggerFactory.getLogger(FlinkDistribution.class);
+
+	private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+	private final List<AutoClosablePath> filesToDelete = new ArrayList<>(4);
+
+	private static final Path FLINK_CONF_YAML = Paths.get("flink-conf.yaml");
+	private static final Path FLINK_CONF_YAML_BACKUP = Paths.get("flink-conf.yaml.bak");
+
+	private final Path opt;
+	private final Path lib;
+	private final Path conf;
+	private final Path log;
+	private final Path bin;
+
+	private Configuration defaultConfig;
+
+	public FlinkDistribution() {
+		final String distDirProperty = System.getProperty("distDir");
+		if (distDirProperty == null) {
+			Assert.fail("The distDir property was not set. You can set it when running maven via -DdistDir=<path> .");
+		}
+		final Path flinkDir = Paths.get(distDirProperty);
+		bin = flinkDir.resolve("bin");
+		opt = flinkDir.resolve("opt");
+		lib = flinkDir.resolve("lib");
+		conf = flinkDir.resolve("conf");
+		log = flinkDir.resolve("log");
+	}
+
+	@Override
+	protected void before() throws IOException {
+		defaultConfig = new UnmodifiableConfiguration(GlobalConfiguration.loadConfiguration(conf.toAbsolutePath().toString()));
+		final Path originalConfig = conf.resolve(FLINK_CONF_YAML);
+		final Path backupConfig = conf.resolve(FLINK_CONF_YAML_BACKUP);
+		Files.copy(originalConfig, backupConfig);
+		filesToDelete.add(new AutoClosablePath(backupConfig));
+	}
+
+	@Override
+	protected void after() {
+		try {
+			stopFlinkCluster();
+		} catch (IOException e) {
+			LOG.error("Failure while shutting down Flink cluster.", e);
+		}
+
+		final Path originalConfig = conf.resolve(FLINK_CONF_YAML);
+		final Path backupConfig = conf.resolve(FLINK_CONF_YAML_BACKUP);
+
+		try {
+			Files.move(backupConfig, originalConfig, StandardCopyOption.REPLACE_EXISTING);
+		} catch (IOException e) {
+			LOG.error("Failed to restore flink-conf.yaml", e);
+		}
+
+		for (AutoCloseable fileToDelete : filesToDelete) {
+			try {
+				fileToDelete.close();
+			} catch (Exception e) {
+				LOG.error("Failure while cleaning up file.", e);
+			}
+		}
+	}
+
+	public void startFlinkCluster() throws IOException {
+		AutoClosableProcess.runBlocking("Start Flink cluster", bin.resolve("start-cluster.sh").toAbsolutePath().toString());
+
+		final OkHttpClient client = new OkHttpClient();
+
+		final Request request = new Request.Builder()
+			.get()
+			.url("http://localhost:8081/taskmanagers")
+			.build();
+
+		Exception reportedException = null;
+		for (int x = 0; x < 30; x++) {
+			try (Response response = client.newCall(request).execute()) {
```

Comment:
Should we wait here for all Task Managers to be up before reporting that the cluster is up and running? This can also be a future JIRA though.

Reply:
This behavior is in sync with the current bash implementation. start-cluster by default only starts 1 TM, so as it stands this is sufficient to guarantee a fully started cluster.

If more TMs are desired we can extend the class to support calls to `taskmanager.sh start` later on.

If we want to support starting a cluster with as many TMs as desired with a single call we'll first have to look at how we could determine how many TMs we should wait for.